Add customization to the scoreboard

J4yTr1n1ty opened the pull request:
This pull request is adding some features I've wanted to use myself and/or that had been asked if it was possible in the discord channels.

## Main Additions
- Customizable Colors for attack and defense to suit branding
- Customizable Sponsor Images and making it rotate properly

### Customizable Colors
Here I am thinking of two different functionalities, one is being able to set defaults in the JSON configuration file already existing in the previous version since #4. Additionally, the user of the HUD should be able to customize their HUD using query parameters,  like the group code already does.

The colors are applied to the following elements:
- Background of Top Info
- Background of Buy Phase Scoreboard
- Background Gradient Start of Agent Pictures
- Background of Shield icon and text
- Background of Weapon Icons

### Customizable Sponsor Images
Since this is already working with a single image, I have already expanded this (in 68cbd2d1f8429e7becc80f1c3f8b01aac588e3c5) to add multiple images in the config, since this is most likely only used in self-hosted scenarios. If there's enough need for it, I might add this to query parameters too, however this would make them very crowded which is not great for readability of links.

## Considerations/Ideas for the future
Due to how dependent on users these configurations are, it would be possible to make them be passed by the client. However, I would refrain from doing that since the client already handles a lot for "only" being an observer client. In my opinion, an Admin Panel or similar should be added in the future to customize the HUD before observers can simply launch their client, type in their authentication data and launch the game. A tournament admin, or in some cases also the same person, can then submit all the information that's needed for the HUD to be customized correctly beforehand. This could be temporarily stored in something like a Redis database or even memory (scalability would be a problem here though), so it can easily be retrieved.

J4yTr1n1ty commented:
Some in between results of how the colors are going to be changed. Currently working on getting the SVGs colored by the css variables. You can also see the sponsors that can be added in the config file.
![image](https://github.com/user-attachments/assets/391c7f34-1a50-4c21-b8a7-c1b719095119)

b4iterdev commented:
Great works !

DarkView commented:
Heads up, in https://github.com/ValoSpectra/Spectra-Frontend/commit/3ffb1601e8baca385f2780a4b5feb3c10c1c8cad we changed the naming of our logo a bit - icon will refer to the 1:1 image of the specter, while logo refers to the icon with text

J4yTr1n1ty commented:
Thank you, rebased and updated the logos to match.

Meanwhile, I've been hitting some issues with how the SVG coloring is going to work. Currently, what I'm doing is updating the SVG with a function and replacing the fill color. 
https://github.com/ValoSpectra/Spectra-Frontend/blob/b889f2ee8b09cd9aca380b8d48f440e290c7942d/src/app/combat/playercard/playercard.component.ts#L99-L112

However, this has some issues because currently, the Input for the component will not trigger change detection and therefore not re-trigger the updateSvg method in, for example, the testing component. This would have to be changed to a BehaviorSubject so we can track the changes. If there's any other ideas you have to solve this, I'm very open to suggestions as I don't want to just rewrite a lot of the main functionality of the application without your opinions first @DarkView @Tiranthine.

DarkView commented:
Yeah we tried changing the color with CSS and the like too, so we don't have to have different files for each color.
However, as you can tell by the presence of different files for each color, we couldn't figure it out sadly.
We'll just have to try different things out too, as using a script to replace the colors is... suboptimal.

J4yTr1n1ty commented:
> as using a script to replace the colors is... suboptimal.

Definitely true, the simple other option is to just have a separate component for the color. That would solve the problem without having to rewrite everything to use Subjects. I'll try to implement it that way and then you can see if that's something that makes sense to you. I haven't found any other way to change the pictures other than using something like an SVG overlay over a white version of the graphic or a calculated filter.

DarkView commented:
Yeah that was the only way we found too.
It's definitely weird.

J4yTr1n1ty commented:
![image](https://github.com/user-attachments/assets/b51b1feb-4271-4547-8f51-5a8b9c20c97c)
Doing everything in a separate component will work. So technically we can leave it like this for now. Definitely would have to think of changing the architecture a bit to better make use of BehaviorSubjects or signals.

J4yTr1n1ty commented:
Since it's probably rarely gonna happen that changes need to be made to the Shield Icons such as recently, I think this option will hold out for a long time while keeping it all working with custom colors.

J4yTr1n1ty commented:
A final screenshot of the Scoreboard as well
![image](https://github.com/user-attachments/assets/5e719073-b197-4b3d-a686-73889b0402e6)

J4yTr1n1ty commented:
@b4iterdev If you can recreate this in html/CSS, you're most welcome to PR it. Might test around with some changes in a separate branch, but not intending to add this in this PR.

DarkView commented:
@b4iterdev that is a 1:1 copy of the official overlay, we will outright deny any change that is this close to the official overlay.
We want to keep a cordial relationship with Riot, copying their design this closely is an easy way to get into trouble.
We can't police what you do in a self-host but we cannot allow something like this in Spectra itself.

b4iterdev commented:
> @b4iterdev that is a 1:1 copy of the official overlay, we will outright deny any change that is this close to the official overlay. We want to keep a cordial relationship with Riot, copying their design this closely is an easy way to get into trouble. We can't police what you do in a self-host but we cannot allow something like this in Spectra itself.

Ik but this is cool tho, also I don't think Riot would be mad at it as there are 3rd party overlay exist having almost similar or exact of "Barracks but for Chinese" (they use a custom overlay based on Barracks but not ran by them afaik). 

This would be just a idea that just came out of my mind, so it's reasonable if it's not acceptable in Spectra overlay

J4yTr1n1ty commented:
Successfully rebased everything onto the current main branch and ran formatting as needed.

J4yTr1n1ty commented:
This currently changes the default color in the config file here https://github.com/ValoSpectra/Spectra-Frontend/blob/06e9cd2bb4e2e204ad4ce992829e07fddec79932/public/assets/config/config.json#L6

for easy review if everything is still recolored as it should. This should be changed before merging or immediately after so people that don't want the custom color don't get forced onto it

DarkView commented:
Looking at the colors side by side, sadly the calculation approach to lightening the colors seems to produce results that are noticeably off from the original.
We could have users provide the lightened color as well and leave the whole color theory thing to them.
We could also re-introduce the multi-step-gradient by having users optionally provide all the needed steps themselves.

DarkView commented on README.md:
```diff
@@ -32,17 +35,24 @@ services:
+  "sponsorImageUrls": ["/assets/misc/logo.webp"],
+  "sponsorImageRotateSpeed": 5000,
+  "attackerColor": "#b82e3c",
+  "defenderColor": "#3c82b8"
```
Might be worth leaving a reference to the config.ts file here to show people an always up-to-date list of available options?

J4yTr1n1ty replied:
resolved and updated to the most current values in e8bc02d26f3be8a90911377ce7eb29e55e37b4ba

DarkView commented:
Looks good to me otherwise!
We'll merge this in alongside #11 for a larger update after the holidays :)

DarkView commented on src/app/combat/playercard/playercard.component.scss:
```diff
@@ -180,17 +180,25 @@
-    border-color: #ff4557;
-    background: linear-gradient(90deg, rgba(255, 69, 87, 0.3) 0%, rgba(31, 122, 88, 0) 100%),
+    border-color: var(--attacker-color);
+    background: linear-gradient(
```
Same for the defender side obviously

J4yTr1n1ty replied:
resolved in e8bc02d26f3be8a90911377ce7eb29e55e37b4ba

J4yTr1n1ty commented:
Applied the requested changes. Happy holidays!
